doc/sphinx/proof-engine/ltac.rst

```diff
@@ -1160,6 +1173,388 @@ Printing |Ltac| tactics
+can now prove lemmas as follows:
+
+.. coqtop:: in
+
+   Lemma solve_perm_ex1 :
+     perm nat (1 :: 2 :: 3 :: nil) (3 :: 2 :: 1 :: nil).
+   Proof. solve_perm. Qed.
+
+.. coqtop:: in
+
+   Lemma solve_perm_ex2 :
+     perm nat
+       (0 :: 1 :: 2 :: 3 :: 4 :: 5 :: 6 :: 7 :: 8 :: 9 :: nil)
+         (0 :: 2 :: 4 :: 6 :: 8 :: 9 :: 7 :: 5 :: 3 :: 1 :: nil).
+   Proof. solve_perm. Qed.
+
+Deciding intuitionistic propositional logic
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Pattern matching on goals allows a powerful backtracking when returning tactic
+values. An interesting application is the problem of deciding intuitionistic
+propositional logic. Considering the contraction-free sequent calculi LJT* of
+Roy Dyckhoff :cite:`Dyc92`, it is quite natural to code such a tactic using the
+tactic language as shown below.
+
+.. coqtop:: in reset
+
+   Ltac basic :=
+   match goal with
+       | |- True => trivial
+       | _ : False |- _ => contradiction
+       | _ : ?A |- ?A => assumption
+   end.
+
+.. coqtop:: in
+
+   Ltac simplify :=
+   repeat (intros;
+       match goal with
+           | H : ~ _ |- _ => red in H
+           | H : _ /\ _ |- _ =>
+               elim H; do 2 intro; clear H
+           | H : _ \/ _ |- _ =>
+               elim H; intro; clear H
+           | H : ?A /\ ?B -> ?C |- _ =>
+               cut (A -> B -> C);
+                   [ intro | intros; apply H; split; assumption ]
+           | H: ?A \/ ?B -> ?C |- _ =>
+               cut (B -> C);
+                   [ cut (A -> C);
+                       [ intros; clear H
+                       | intro; apply H; left; assumption ]
+                   | intro; apply H; right; assumption ]
+           | H0 : ?A -> ?B, H1 : ?A |- _ =>
+               cut B; [ intro; clear H0 | apply H0; assumption ]
+           | |- _ /\ _ => split
+           | |- ~ _ => red
+       end).
+
+.. coqtop:: in
+
+   Ltac my_tauto :=
+     simplify; basic ||
+     match goal with
+         | H : (?A -> ?B) -> ?C |- _ =>
+             cut (B -> C);
+                 [ intro; cut (A -> B);
+                     [ intro; cut C;
+                         [ intro; clear H | apply H; assumption ]
+                     | clear H ]
+                 | intro; apply H; intro; assumption ]; my_tauto
+         | H : ~ ?A -> ?B |- _ =>
+             cut (False -> B);
+                 [ intro; cut (A -> False);
+                     [ intro; cut B;
+                         [ intro; clear H | apply H; assumption ]
+                     | clear H ]
+                 | intro; apply H; red; intro; assumption ]; my_tauto
+         | |- _ \/ _ => (left; my_tauto) || (right; my_tauto)
+     end.
+
+The tactic ``basic`` tries to reason using simple rules involving truth, falsity
+and available assumptions. The tactic ``simplify`` applies all the reversible
+rules of Dyckhoff’s system. Finally, the tactic ``my_tauto`` (the main
+tactic to be called) simplifies with ``simplify``, tries to conclude with
```

Comment:
"The tactic ``simplify`` applies all the reversible rules of Dyckhoff’s system." needs some unpacking, at least for me.  Does simplify do this out of the box?